contracts/src/L2/L2BridgeController.sol

```diff
@@ -0,0 +1,216 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.13;
+
+import {TransferData, MigrationData} from "../common/TransferData.sol";
+
+import {IRegistry} from "../common/IRegistry.sol";
+import {IRegistryDatastore} from "../common/IRegistryDatastore.sol";
+import {NameCoder} from "@ens/contracts/utils/NameCoder.sol";
+import {NameUtils} from "../common/NameUtils.sol";
+import {IERC1155Receiver} from "@openzeppelin/contracts/token/ERC1155/IERC1155Receiver.sol";
+import {ERC165} from "@openzeppelin/contracts/utils/introspection/ERC165.sol";
+import {IERC165} from "@openzeppelin/contracts/utils/introspection/IERC165.sol";
+import {ITokenObserver} from "../common/ITokenObserver.sol";
+import {IPermissionedRegistry} from "../common/IPermissionedRegistry.sol";
+import {EjectionController} from "../common/EjectionController.sol";
+import {IBridge} from "../common/IBridge.sol";
+import {BridgeEncoder} from "../common/BridgeEncoder.sol";
+import {LibEACBaseRoles} from "../common/EnhancedAccessControl.sol";
+import {IEnhancedAccessControl} from "../common/IEnhancedAccessControl.sol";
+import {RegistryRolesMixin} from "../common/RegistryRolesMixin.sol";
+
+/**
+ * @title L2BridgeController
+ * @dev Combined controller that handles both migration messages from L1 to L2 and ejection operations
+ */
+contract L2BridgeController is EjectionController, ITokenObserver, RegistryRolesMixin {
+    error MigrationFailed();
+    error InvalidTLD(bytes dnsEncodedName);
+    error NameNotFound(bytes dnsEncodedName);
+    error NotTokenOwner(uint256 tokenId);
+    error TooManyRoleAssignees(uint256 tokenId, uint256 roleBitmap);
+
+    // Events
+    event MigrationCompleted(bytes dnsEncodedName, uint256 newTokenId);
+
+    bytes32 public constant ETH_TLD_HASH = keccak256(bytes("eth"));
+
+    IRegistryDatastore public immutable datastore;
+
+    constructor(
+        IBridge _bridge,
+        IPermissionedRegistry _registry, 
+        IRegistryDatastore _datastore
+    ) EjectionController(_registry, _bridge) {
+        datastore = _datastore;
+    }   
+
+    /**
+     * @dev Complete migration from L1 to L2
+     * Called by the bridge when a migration message is received from L1
+     * 
+     * @param dnsEncodedName The DNS encoded name being migrated
+     * @param migrationData The migration data containing transfer details
+     */
+    function completeMigrationFromL1(
+        bytes memory dnsEncodedName,
+        MigrationData memory migrationData
+    ) external onlyBridge {
+        // if migrating to L1 then there is nothing to do, else let's create a subregistry
+        if (!migrationData.toL1) {
```

Comment:
Is this field and this `if` statement still required?

Reply:
Without the `if`, it looks like the migration and ejection code might actually be identical?

Reply:
You're right, will refactor.